src/audio/pcm_converter/pcm_converter_hifi3.c

```diff
@@ -462,70 +328,49 @@ static int pcm_convert_s24_to_s32(const struct audio_stream __sparse_cache *sour
 				  uint32_t ioffset, struct audio_stream __sparse_cache *sink,
 				  uint32_t ooffset, uint32_t samples)
 {
-	ae_int32x2 *in = audio_stream_read_frag(source, ioffset,
-						sizeof(int32_t));
-	ae_int32x2 *out = audio_stream_write_frag(sink, ooffset,
-						  sizeof(int32_t));
+	int32_t *src = source->r_ptr;
+	int32_t *dst = sink->w_ptr;
 	ae_int32x2 sample = AE_ZERO32();
-	ae_valign align_out = AE_ZALIGN64();
-	int i;
+	int nmax, i, n, m;
+	ae_valign outu = AE_ZALIGN64();
+	ae_valign inu = AE_ZALIGN64();
+	int left_samples = samples;
+	ae_int32x2 *in = audio_stream_wrap(source, src + ioffset);
+	ae_int32x2 *out = audio_stream_wrap(sink, dst + ooffset);
 
 	/* nothing to do */
 	if (!samples)
 		return samples;
```

Comment:
in all the code - this is redundant if we keep the `while ()` loop. You can even do the loop as `for (left_samples = samples; left_samples; )`. In any case I think we should optimise this out - either remove this check or modify the loop

Reply:
Yes, you are right. I am waiting the failed case info to analyse if there is anything else to be correct. Will modify as suggested next.